src/lib/y2storage/hwinfo_reader.rb

```diff
@@ -0,0 +1,120 @@
+# encoding: utf-8
+
+# Copyright (c) [2017] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "ostruct"
+require "singleton"
+
+module Y2Storage
+  # Read hardware information for storage devices
+  #
+  # This class uses `hwinfo` to read hardware information from storage devices.
+  # The information is read lazily for all devices at once and cached when the
+  # #for_device method is called.
+  #
+  # The cache can be cleaned calling #reset.
+  #
+  # @example Get information for "/dev/sda"
+  #   hwinfo = HWInfo.instance.for_device(name)
+  #   hwinfo.driver #=> ["ahci", "sd"]
+  #   hwinfo.device_file #=> "/dev/sda"
+  #
+  # @example Cleaning the cache
+  #   HWInfo.instance.reset!
+  class HWInfoReader
+    include Singleton
+
+    # Return hardware info for the given device
+    #
+    # @param name [String] Device name (eg. "/dev/sda")
+    # @return [OpenStruct] Hardware information
+    def for_device(name)
+      data[name]
+    end
+
+    # Reset the cache
+    #
+    # The values will be loaded when #for_device is called again.
+    def reset!
```

Comment:
Lately, we are only using bang methods when we have two versions, with and without exclamation mark.